### PR TITLE
[SLA] PIM-5276: Fixed attribute ordering in the product view

### PR DESCRIPTION
**Pull request by ClementGautier**
| Q                 | A
| ----------------- | ---
| Specs             | N/A
| Behats            | Yes
| Blue CI           | Running
| Changelog updated | Y 
| Review and 2 GTM  | todo

### CODE REVIEW

File: features/Context/Page/AttributeGroup/Edit.php [hunk 1]
```diff
@@ -36,23 +36,30 @@ public function __construct($session, $pageFactory, $parameters = array())
...
-        $list = $this->getElement('Attribute list')->findAll('css', 'tr');
-        $elt  = $this->getElement('Attribute list')->find('css', sprintf('tr:contains("%s")', $attribute));
+        $selector             = $this->getElement('Attribute list')->getAttribute('id');
+        $currentPosition      = $this->getAttributePosition($attribute);
+        $manipulationFunction = 'insertAfter';
+        $position--;
 
-        if (!$elt) {
-            throw new \InvalidArgumentException(sprintf('Attribute %s was not found', $attribute));
+        if (0 == $position) {
```
Comment (solivier) [inline features/Context/Page/AttributeGroup/Edit.php]: === ?

File: features/product/order_attributes.feature [hunk 2]
```diff
@@ -6,30 +6,38 @@ Feature: Order attributes
...
+    When I am on the "info" attribute group page
+    And I visit the "Attributes" tab
+    And I change the attribute "Description" position to 2
+    And I am on the "boot-001" product page
+    Then the attribute "SKU" should be in position 1
+    And the attribute "Description" should be in position 2
+    And the attribute "Name" should be in position 3
+    And the attribute "Manufacturer" should be in position 4
+    And the attribute "Weather conditions" should be in position 5
+    And the attribute "Length" should be in position 6
```
Comment (solivier) [inline features/product/order_attributes.feature]: missing new line

File: spec/Pim/Bundle/VersioningBundle/Doctrine/MongoDBODM/BulkVersionBuilderSpec.php [hunk 3]
```diff
@@ -14,6 +14,12 @@
 use Prophecy\Argument;
 use Symfony\Component\EventDispatcher\EventDispatcherInterface;
 
+/**
+ * @require Doctrine\MongoDB\Collection
```
Comment (juliensnz) [inline spec/Pim/Bundle/VersioningBundle/Doctrine/MongoDBODM/BulkVersionBuilderSpec.php]: what ?

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/AttributeRepository.php [hunk 4]
```diff
@@ -476,6 +476,7 @@ public function getAttributeCodesByGroup(AttributeGroupInterface $group)
         $qb
             ->select('a.code')
             ->where($qb->expr()->eq('a.group', ':group'))
+            ->orderBy('a.sortOrder')
```
Comment (juliensnz) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/AttributeRepository.php]: it's not safe and you should do that on the front end

Comment (juliensnz) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/AttributeRepository.php]: you should keep that an sort them on the frontend too

File: features/Context/Page/AttributeGroup/Edit.php [hunk 5]
```diff
@@ -36,23 +36,30 @@ public function __construct($session, $pageFactory, $parameters = array())
...
-            throw new \InvalidArgumentException(
-                sprintf('Unable to change the position to %d, only %s attributes present', $position, count($list))
-            );
-        }
-
-        $eltHandle = $elt->find('css', '.handle');
-        $target    = $list[$position-1];
+        // Move the row
+        $this->getDriver()->evaluateScript(sprintf(
+            'console.log($("#%s tr:eq(%d)").%s( $( "#attributes-sortable tr:eq(%d)" ) ));',
```
Comment (juliensnz) [inline features/Context/Page/AttributeGroup/Edit.php]: please remove that

File: features/product/order_attributes.feature (hunk 2 above)
Comment (juliensnz) [inline features/product/order_attributes.feature]: add a blank line at the end

File: features/product/order_attributes.feature [hunk 6]
```diff
@@ -6,30 +6,38 @@ Feature: Order attributes
...
+  Scenario: Update an attribute position in the product view
+    Given the following product:
+      | sku      | description-en_US-mobile | name-en_US | manufacturer | weather_conditions | length        |
+      | boot-001 | Nice dark rangers        | Rangers    | Caterpillar  | Wet                | 50 CENTIMETER |
+    When I am on the "info" attribute group page
+    And I visit the "Attributes" tab
+    And I change the attribute "Description" position to 2
+    And I am on the "boot-001" product page
+    Then the attribute "SKU" should be in position 1
+    And the attribute "Description" should be in position 2
```
Comment (juliensnz) [inline features/product/order_attributes.feature]: this is already the normal order

File: features/Context/Page/Product/Edit.php [hunk 7]
```diff
@@ -443,6 +443,24 @@ public function findFieldContainer($label)
...
+    {
+        $rows = $this->find('css', '.tab-pane.active.product-values')->findAll('css', 'div.field-container');
+
+        foreach ($rows as $index => $row) {
+            if ($row->find('css', sprintf(':contains("%s")', $attribute))) {
+                return $index+1;
+            }
+        }
+
+        throw new \InvalidArgumentException(sprintf('Attribute %s was not found', $attribute));
```
Comment (skeleton) [inline features/Context/Page/Product/Edit.php]: ElementNotFoundException ?

**Comment by solivier** [comment]
GTM after comments fixed

File: features/Context/Page/AttributeGroup/Edit.php (hunk 5 above)
Comment (skeleton) [inline features/Context/Page/AttributeGroup/Edit.php]: unreadable :(

File: spec/Pim/Bundle/VersioningBundle/Doctrine/MongoDBODM/BulkVersionBuilderSpec.php (hunk 3 above)
Comment (phaseinducer) [inline spec/Pim/Bundle/VersioningBundle/Doctrine/MongoDBODM/BulkVersionBuilderSpec.php]: Yep sorry, I broke the specs with https://github.com/akeneo/pim-community-dev/pull/3683

File: features/Context/Page/Product/Edit.php [hunk 8]
```diff
@@ -443,6 +443,24 @@ public function findFieldContainer($label)
     }
 
     /**
+     * @param string $attribute
+     *
+     * @return int
+     */
+    public function getAttributePosition($attribute)
+    {
+        $rows = $this->find('css', '.tab-pane.active.product-values')->findAll('css', 'div.field-container');
```
Comment (phaseinducer) [inline features/Context/Page/Product/Edit.php]: `div` is useless

File: features/Context/Page/Product/Edit.php [hunk 9]
```diff
@@ -443,6 +443,24 @@ public function findFieldContainer($label)
...
+
+        foreach ($rows as $index => $row) {
+            if ($row->find('css', sprintf(':contains("%s")', $attribute))) {
+                return $index+1;
```
Comment (phaseinducer) [inline features/Context/Page/Product/Edit.php]: spaces around + please

**Comment by phaseinducer** [comment]
GTM after fix

File: features/product/order_attributes.feature [hunk 10]
```diff
@@ -6,30 +6,50 @@ Feature: Order attributes
...
+    And the attribute "Manufacturer" should be in position 3
+    And the attribute "Weather conditions" should be in position 4
+    And the attribute "Length" should be in position 5
+    When I display the Name attribute
+    Then the attribute "SKU" should be in position 1
+    And the attribute "Description" should be in position 2
+    And the attribute "Name" should be in position 3
+    And the attribute "Manufacturer" should be in position 4
+    And the attribute "Weather conditions" should be in position 5
+    And the attribute "Length" should be in position 6
```
Comment (juliensnz) [inline features/product/order_attributes.feature]: could you add a before/after ?

File: features/product/order_attributes.feature [hunk 11]
```diff
@@ -6,30 +6,50 @@ Feature: Order attributes
...
+    When I am on the "info" attribute group page
+    And I visit the "Attributes" tab
+    And I change the attribute "Description" position to 2
+    And I am on the "boot-001" product page
+    Then the attribute "SKU" should be in position 1
+    And the attribute "Description" should be in position 2
+    And the attribute "Manufacturer" should be in position 3
+    And the attribute "Weather conditions" should be in position 4
+    And the attribute "Length" should be in position 5
+    When I display the Name attribute
```
Comment (juliensnz) [inline features/product/order_attributes.feature]: is it the way to add an optional attribute ?

**Comment by juliensnz** [comment]
GTM, good job !

**Pull request merged**